test/e2e/cluster_upgrade_test.go

```diff
@@ -69,9 +83,115 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA
 			// This test is run in CI in parallel with other tests. To keep the test duration reasonable
 			// the conformance tests are skipped.
 			SkipConformanceTests:     true,
-			ControlPlaneMachineCount: ptr.To[int64](3),
+			ControlPlaneMachineCount: ptr.To[int64](controlPlaneMachineCount),
 			WorkerMachineCount:       ptr.To[int64](1),
-			Flavor:                   ptr.To("topology"),
+			Flavor:                   ptr.To("kcp-pre-drain"),
+			PreWaitForControlPlaneToBeUpgraded: func(managementClusterProxy framework.ClusterProxy, workloadClusterNamespace, workloadClusterName string) {
+				log.Logf("Waiting for control-plane machines to have the upgraded Kubernetes version")
+
+				preDrainHook := "pre-drain.delete.hook.machine.cluster.x-k8s.io/kcp-ready-check"
+
+				cluster := &clusterv1.Cluster{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: workloadClusterNamespace,
+						Name:      workloadClusterName,
+					},
+				}
+				Expect(managementClusterProxy.GetClient().Get(ctx, client.ObjectKeyFromObject(cluster), cluster)).To(Succeed())
+
+				// This replaces the WaitForControlPlaneMachinesToBeUpgraded function to ensure that
+				// all static Pods and kube-proxy as well as all Nodes are healthy during the whole upgrade
+				// across all Nodes of the Cluster. For doing this it uses a pre-drain hook.
```

Comment:
```suggestion

				// This replaces the WaitForControlPlaneMachinesToBeUpgraded function to verify via a pre-drain hook 
				// that all static Pods, kube-proxy and all Nodes are becoming healthy before we let the upgrade 
				// process precede by removing the pre-drain hook.
				// This captures cases where static Pods, kube-proxy and all Nodes would only become healthy after 
				// all control plane Machines have been upgraded.
```
I know it's just a nuance but I think it's an important one